cmd/object-handlers.go

```diff
@@ -689,6 +689,185 @@ func (api objectAPIHandlers) getObjectHandler(ctx context.Context, objectAPI Obj
+
+	if _, ok := opts.ObjectAttributes[xhttp.StorageClass]; ok {
+		OA.StorageClass = objInfo.StorageClass
+	}
+
+	err = objInfo.decryptPartsChecksums()
+	if err != nil {
+		writeErrorResponseHeadersOnly(w, toAPIError(ctx, err))
+		return
+	}
+
+	if _, ok := opts.ObjectAttributes[xhttp.ObjectParts]; ok {
+		OA.ObjectParts = new(objectAttributesParts)
+		OA.ObjectParts.PartNumberMarker = opts.PartNumberMarker
+
+		OA.ObjectParts.MaxParts = opts.MaxParts
+		partsLength := len(objInfo.Parts)
+		OA.ObjectParts.PartsCount = partsLength
+
+		if opts.MaxParts > -1 {
+			for i, v := range objInfo.Parts {
+				if v.Number <= opts.PartNumberMarker {
+					continue
+				}
+
+				if len(OA.ObjectParts.Parts) == opts.MaxParts {
+					break
+				}
+
+				OA.ObjectParts.NextPartNumberMarker = v.Number
+				OA.ObjectParts.Parts = append(OA.ObjectParts.Parts, &objectAttributesPart{
+					ChecksumSHA1:   objInfo.Parts[i].Checksums["SHA1"],
+					ChecksumSHA256: objInfo.Parts[i].Checksums["SHA256"],
+					ChecksumCRC32:  objInfo.Parts[i].Checksums["CRC32"],
+					ChecksumCRC32C: objInfo.Parts[i].Checksums["CRC32C"],
+					PartNumber:     objInfo.Parts[i].Number,
+					Size:           objInfo.Parts[i].Size,
+				})
+			}
+		}
+
+		if OA.ObjectParts.NextPartNumberMarker != partsLength {
+			OA.ObjectParts.IsTruncated = true
+		}
+	}
+
+	_, _ = w.Write([]byte(`<?xml version="1.0" encoding="UTF-8"?>`))
+	if err := xml.NewEncoder(w).Encode(OA); err != nil {
+		writeErrorResponseHeadersOnly(w, toAPIError(ctx, err))
+		return
+	}
+
+	sendEvent(eventArgs{
+		EventName:    event.ObjectAccessedAttributes,
+		BucketName:   bucket,
+		Object:       objInfo,
+		ReqParams:    extractReqParams(r),
+		RespElements: extractRespElements(w),
+		UserAgent:    r.UserAgent(),
+		Host:         handlers.GetSourceIP(r),
+	})
+
+	return
+}
+
+func detectAndHandleNotFoundError(ctx context.Context, r *http.Request, w http.ResponseWriter, bucket, object, versionID string, err error) (hasError bool) {
+	var outErr error
+	if isErrObjectNotFound(err) {
+		outErr = ObjectNotFound{
+			Bucket:    bucket,
+			Object:    object,
+			VersionID: versionID,
+			Err:       err,
+		}
+		hasError = true
+	} else if isErrVersionNotFound(err) {
+		outErr = VersionNotFound{
+			Bucket:    bucket,
+			Object:    object,
+			VersionID: versionID,
+			Err:       err,
+		}
+		hasError = true
+	}
+
+	if !hasError {
+		return false
+	}
+
+	reqInfo := logger.GetReqInfo(ctx)
+
+	if reqInfo == nil {
+		writeErrorResponse(ctx, w, errorCodes.ToAPIErr(ErrAccessDenied), r.URL)
+		return true
+	}
+
+	if globalIAMSys.IsAllowed(policy.Args{
+		AccountName:     reqInfo.Cred.AccessKey,
+		Groups:          reqInfo.Cred.Groups,
+		Action:          policy.ListBucketAction,
+		BucketName:      bucket,
+		ConditionValues: getConditionValues(r, "", reqInfo.Cred),
+		ObjectName:      object,
+		IsOwner:         reqInfo.Owner,
+		Claims:          reqInfo.Cred.Claims,
+	}) {
+		writeErrorResponse(ctx, w, toAPIError(ctx, outErr), r.URL)
```

Comment:
When the request is allowed we are writing an error to the client which is empty?

Reply:
This method is used to detect if we should return a 404 or a 403 depending on if the user has s3:ListBucket permissions. But before that check is done the method will actually check if there is an error at all. 

The current functionality works as intended, however.. since it's hard to process the logic flow at a quick glance I did a re-write. Will push the changes soon.